File: Core/Core.Utilities/Saver.vb
Public Class Saver
    Public Sub Save(ByVal transactionHandler As Framework.ITransactionHandler, ByVal save As Action(Of ITransactionHandler))
        Try
            save.Invoke(transactionHandler)

            If transactionHandler.DbTransaction IsNot Nothing Then
                transactionHandler.DbTransaction.Commit()
            End If
        Catch
            If transactionHandler.DbTransaction IsNot Nothing Then
                transactionHandler.DbTransaction.Rollback()
            End If
            Throw
        Finally
            If transactionHandler.DbTransaction IsNot Nothing Then
                transactionHandler.DbTransaction.Dispose()
                transactionHandler.DbTransaction = Nothing
            End If
            If transactionHandler.DbConnection IsNot Nothing Then
                transactionHandler.DbConnection.Dispose()
                transactionHandler.DbConnection = Nothing
            End If
        End Try
    End Sub
End Class

File: API/CommonCore/Saver.cs
using System;
using System.Threading.Tasks;

namespace JestersCreditUnion.CommonCore
{
    public static class Saver
    {
        public static async Task Save(ITransactionHandler transactionHandler, Func<ITransactionHandler, Task> save)
        {
            try
            {
                await save(transactionHandler);
                if (transactionHandler.Transaction != null)
                    transactionHandler.Transaction.Commit();
                if (transactionHandler.Connection != null)
                    transactionHandler.Connection.Close();
            }
            catch
            {
                if (transactionHandler.Transaction != null)
                    transactionHandler.Transaction.Rollback();
                throw;
            }
            finally
            {
                if (transactionHandler.Transaction != null)
                {
                    transactionHandler.Transaction.Dispose();
                    transactionHandler.Transaction = null;
                }
                if (transactionHandler.Connection != null)
                {
                    transactionHandler.Connection.Dispose();
                    transactionHandler.Connection = null;
                }

            }
        }
    }
}
